### PR TITLE
Extend ART to feature vectors

### LINKED ISSUE #49: Adapt ART to feature vectors
This requires multiple changes to all modules:

- [x] Allow `clip_values` to be `None` in all classifiers
- [x] Allow `clip_values` to have a different value per feature
- [ ] Similarly, change attack strength (often `eps`, `eps_step`) from just scalar to scalar or vector. If  a vector is provided, it should have the same size as the number of features. This is to allow attacks to be applied on features with different ranges and mostly concerns L_inf attacks.
- [ ] In line with the previous item, adapt random initialization in boundary attack to take into account the different scales in features for feature vectors.
- [x] Only perform clipping in attacks if the targeted model has `clip_values`
- [x] Generalize the shape of the input assumed in `Classifier` and other modules: moving from four dimensions (first one being the batch) to any number of dimensions (where first one is still the batch size)
- [x] Add checks and safeguards on number of dimensions of data for the attacks / defences that can only be applied on images
- [x] Add tests for feature vectors classifiers
- [x] Add tests with no clipping / clipping per feature for both images and feature vectors
- [ ] Update examples and notebooks to ensure that they still work
- [ ] Create notebook with attacks on feature vectors using some malware dataset

### PR DESCRIPTION
**Pull request by ririnicolae**
# Description
This PR extends ART to feature vectors. `clip_values` are now optional in classifiers. Clipping is only performed when these values are present. Constraints for images shapes have been removed. Attacks and defences were adapted to more general data shapes. Some attacks and defences that can only be applied on images now raise an error if a different data shape is provided. All these changes were tested on the Iris dataset with a small neural network model.

Closes #49

## Type of change

Please check all relevant options.

- [ ] Bug fix (non-breaking)
- [X] New feature (non-breaking)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# Testing

- [X] Add trained Iris models for testing and code for loading them
- [X] Add unit tests for attacks, classifiers, wrappers and defences using Iris dataset

# Checklist

- [X] My code follows the style guidelines of this project
- [X] I have performed a self-review of my own code
- [X] I have commented my code
- [X] I have made corresponding changes to the documentation
- [X] My changes generate no new warnings
- [X] I have added tests that prove my fix is effective or that my feature works
- [X] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: art/attacks/virtual_adversarial.py [hunk 1]
```diff
@@ -116,12 +123,12 @@ def _normalize(x):
         :rtype: `np.ndarray`
         """
         tol = 1e-10
-        dims = x.shape
+        # dims = x.shape
```
Comment (beat-buesser) [inline art/attacks/virtual_adversarial.py]: Do we need to keep this commented line?

File: art/attacks/virtual_adversarial.py [hunk 2]
```diff
@@ -116,12 +123,12 @@ def _normalize(x):
...
 
-        x = x.reshape(dims[0], -1)
+        # x = x.reshape(dims[0], -1)
```
Comment (beat-buesser) [inline art/attacks/virtual_adversarial.py]: Do we need to keep this commented line?

File: art/attacks/virtual_adversarial.py [hunk 3]
```diff
@@ -116,12 +123,12 @@ def _normalize(x):
...
         inverse = (np.sum(x**2, axis=1) + tol) ** -.5
         x = x * inverse[:, None]
-        x = np.reshape(x, dims)
+        # x = np.reshape(x, dims)
```
Comment (beat-buesser) [inline art/attacks/virtual_adversarial.py]: Do we need to keep this commented line?

**Comment by ririnicolae** [comment]
@beat-buesser Thanks for the review and for noticing those comments, I have removed them now!

**Pull request merged**